File: backend/tools/lahore_season.py
"""Lahore season detection — mirrors frontend/lib/lahoreSeason.ts (Innovation 3)."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date, datetime

from services.seasonal_intelligence import get_season_profile, is_smog_season as _is_smog

HEATWAVE_C = 42


@dataclass(frozen=True)
class LahoreSeason:
    id: str
    label_en: str
    label_ur: str


def get_lahore_season(when: date | datetime | None = None) -> LahoreSeason:
    """
    Pakistan-specific 4-season calendar for Lahore:
    Winter Smog (Oct–Jan), Spring Dust (Feb–Apr),
    Summer Heatwave (May–Jul), Monsoon (Aug–Sep).
    """
    d = when.date() if isinstance(when, datetime) else (when or date.today())
    month = d.month

    if month in (10, 11, 12, 1):
        profile = get_season_profile("winter_smog")
    elif month in (2, 3, 4):
        profile = get_season_profile("spring_dust")
    elif month in (5, 6, 7):
        profile = get_season_profile("summer_heatwave")
    else:
        profile = get_season_profile("monsoon")

    return LahoreSeason(profile.id, profile.label_en, profile.label_ur)


def is_heatwave(temp_c: float) -> bool:
    return temp_c >= HEATWAVE_C


def is_smog_season(season_id: str) -> bool:
    return _is_smog(season_id)

File: frontend/lib/lahoreSeason.ts
/** Lahore Seasonal Intelligence — Pakistan-specific 4-season calendar (Innovation 3) */

export type LahoreSeasonId =
  | "winter_smog"
  | "spring_dust"
  | "summer_heatwave"
  | "monsoon";

export interface LahoreSeason {
  id: LahoreSeasonId;
  labelEn: string;
  labelUr: string;
}

const HEATWAVE_C = 42;

/** Oct–Jan | Feb–Apr | May–Jul | Aug–Sep */
export function getLahoreSeason(date = new Date()): LahoreSeason {
  const month = date.getMonth() + 1;

  if (month === 10 || month === 11 || month === 12 || month <= 1) {
    return {
      id: "winter_smog",
      labelEn: "Winter smog / sardi",
      labelUr: "Smog / sardi season",
    };
  }
  if (month >= 2 && month <= 4) {
    return {
      id: "spring_dust",
      labelEn: "Spring dust & pollen",
      labelUr: "Spring / beech ka mausam",
    };
  }
  if (month >= 5 && month <= 7) {
    return {
      id: "summer_heatwave",
      labelEn: "Garmi / heatwave",
      labelUr: "Garmi season",
    };
  }
  return {
    id: "monsoon",
    labelEn: "Barsaat / monsoon",
    labelUr: "Barsaat / monsoon",
  };
}

export function isHeatwave(tempC: number): boolean {
  return tempC >= HEATWAVE_C;
}

export function isSmogSeason(seasonId: LahoreSeasonId): boolean {
  return seasonId === "winter_smog";
}
